testscript/exe.go

```diff
@@ -49,63 +51,138 @@ func IgnoreMissedCoverage() {
+		// their profiles to. Before finishing, we'll merge all of those
+		// profiles into the main profile.
+		if coverProfile() != "" {
+			coverdir := filepath.Join(tmpdir, "cover")
+			if err := os.MkdirAll(coverdir, 0777); err != nil {
+				log.Printf("could not set up cover directory: %v", err)
+				return 2
+			}
+			os.Setenv("TESTSCRIPT_COVER_DIR", coverdir)
+			defer func() {
+				if err := finalizeCoverProfile(coverdir); err != nil {
+					log.Printf("cannot merge cover profiles: %v", err)
+					exitCode = 2
+				}
+			}()
+		}
+
 		// We're not in a subcommand.
 		for name := range commands {
 			name := name
+			// Set up this command in the directory we added to $PATH.
+			binfile := filepath.Join(bindir, name)
+			if runtime.GOOS == "windows" {
+				binfile += ".exe"
+			}
+			binpath, err := os.Executable()
+			if err == nil {
+				err = copyBinary(binpath, binfile)
+			}
+			if err != nil {
+				log.Printf("could not set up %s in $PATH: %v", name, err)
+				return 2
+			}
 			scriptCmds[name] = func(ts *TestScript, neg bool, args []string) {
-				path, err := os.Executable()
-				if err != nil {
-					ts.Fatalf("cannot determine path to test binary: %v", err)
-				}
-				id := atomic.AddInt32(&profileId, 1) - 1
-				oldEnvLen := len(ts.env)
-				cprof := coverFilename(id)
-				ts.env = append(ts.env,
-					"TESTSCRIPT_COMMAND="+name,
-					"TESTSCRIPT_COVERPROFILE="+cprof,
-				)
-				ts.cmdExec(neg, append([]string{path}, args...))
-				ts.env = ts.env[0:oldEnvLen]
-				if cprof == "" {
-					return
-				}
-				f, err := os.Open(cprof)
-				if err != nil {
-					if ignoreMissedCoverage {
-						return
-					}
-					ts.Fatalf("command %s (args %q) failed to generate coverage information", name, args)
-					return
-				}
-				coverChan <- f
+				ts.cmdExec(neg, append([]string{name}, args...))
 			}
 		}
 		return m.Run()
 	}
-	mainf := commands[cmdName]
-	if mainf == nil {
-		log.Printf("unknown command name %q", cmdName)
-		return 2
-	}
 	// The command being registered is being invoked, so run it, then exit.
 	os.Args[0] = cmdName
-	cprof := os.Getenv("TESTSCRIPT_COVERPROFILE")
-	if cprof == "" {
+	coverdir := os.Getenv("TESTSCRIPT_COVER_DIR")
+	if coverdir == "" {
 		// No coverage, act as normal.
 		return mainf()
 	}
+
+	// For a command "foo", write ${TESTSCRIPT_COVER_DIR}/foo-${RANDOM}.
+	// Note that we do not use ioutil.TempFile as that creates the file.
+	// In this case, we want to leave it to -test.coverprofile to create the
+	// file, as otherwise we could end up with an empty file.
+	// Later, when merging profiles, trying to merge an empty file would
+	// result in a confusing error.
+	rnd, err := nextRandom()
+	if err != nil {
+		log.Printf("could not obtain random number: %v", err)
+		return 2
+	}
+	cprof := filepath.Join(coverdir, fmt.Sprintf("%s-%x", cmdName, rnd))
 	return runCoverSubcommand(cprof, mainf)
 }
 
+func nextRandom() ([]byte, error) {
+	p := make([]byte, 4)
```

Comment:
32 bits isn't _that_ huge - might as well use 6 or 8 and make it as secure as a normal hash collision.

Reply:
I actually started at 8, but went down a bit as I thought it was unnecessarily long :) I'll go with 6.